src/Framework/Migrations/Controllers/BatchMigrationRunner.php

```diff
@@ -0,0 +1,214 @@
+<?php
+
+namespace Give\Framework\Migrations\Controllers;
+
+use ActionScheduler_Store;
+use Exception;
+use Give\Framework\Database\DB;
+use Give\Framework\Migrations\Contracts\BatchMigration;
+use Give\MigrationLog\MigrationLogStatus;
+
+/**
+ * Batch Migration runner controller
+ *
+ * @unreleased
+ */
+class BatchMigrationRunner
+{
+    /**
+     * @var BatchMigration
+     */
+    private $migration;
+
+    /**
+     * @var ActionScheduler_Store
+     */
+    private $actionSchedulerStore;
+
+    /**
+     * @unreleased
+     */
+    public function __construct(BatchMigration $migration)
+    {
+        $this->migration = $migration;
+        $this->actionSchedulerStore = ActionScheduler_Store::instance();
+    }
+
+    /**
+     * Run batch migration
+     *
+     * @unreleased
+     *
+     * @throws Exception
+     */
+    public function run(): string
+    {
+        $this->registerAction();
+
+        $actions = $this->actionSchedulerStore->query_actions([
+            'group' => $this->getGroup(),
+            'per_page' => 0,
+        ]);
+
+        if ( ! $actions) {
+            $itemsCount = $this->migration->getItemsCount();
+
+            // Bailout if there are no items to process
+            if ( ! $itemsCount) {
+                return MigrationLogStatus::SUCCESS;
+            }
+
+            $current = 0;
+            $batches = ceil($itemsCount / $this->migration->getBatchSize());
+
+            // Register migration action for each batch
+            for ($i = 0; $i < $batches; $i++) {
+                if ($items = $this->migration->getBatchItemsAfter($current)) {
+                    [$firstId, $lastId] = $items;
+                    $current = $lastId;
+                    as_enqueue_async_action($this->getHook(), [$firstId, $lastId], $this->getGroup());
+                }
+            }
+
+            return MigrationLogStatus::RUNNING;
+        }
+
+        $pendingActions = (int)$this->actionSchedulerStore->query_actions([
+            'group' => $this->getGroup(),
+            'status' => [
+                ActionScheduler_Store::STATUS_RUNNING,
+                ActionScheduler_Store::STATUS_PENDING,
+            ],
+        ], 'count');
+
+        if ($pendingActions) {
```

Comment:
Is this the total number of actions running? Should we compare this to the `$batches` number to make sure they're equal?

Same question below for `$failedActions`.